### PR TITLE
Add gadgets for elliptic curve operations

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

File: src/gadget/ecc.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    type Scalar: Clone + fmt::Debug;
+    /// Variable representing an elliptic curve point.
+    type Point: Clone + fmt::Debug;
+    /// Variable representing a fixed elliptic curve point (constant in the circuit).
+    type FixedPoint: Clone + fmt::Debug;
+
+    /// Witnesses the given scalar as a private input to the circuit.
+    fn witness_scalar(
+        layouter: &mut impl Layouter<Self>,
+        value: impl FnMut() -> Result<C::Scalar, Error>,
```
Comment (str4d) [inline src/gadget/ecc.rs:23]: We don't actually want this to be a function; we need to optionally map over the value if it exists, and with a function we can only learn this by calling the function, making it redundant.
```suggestion
        value: Option<C::Scalar>,
```

File: src/gadget/ecc.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    ) -> Result<Self::Scalar, Error>;
+
+    /// Witnesses the given point as a private input to the circuit.
+    fn witness_point(
+        layouter: &mut impl Layouter<Self>,
+        value: impl FnMut() -> Result<C, Error>,
+    ) -> Result<Self::Point, Error>;
+
+    /// Exposes the given point as a public input to the circuit.
+    fn expose_point(layouter: &mut impl Layouter<Self>, a: &Self::Point) -> Result<(), Error>;
```
Comment (str4d) [inline src/gadget/ecc.rs:33]: Sean says let's just remove this for now.

File: src/gadget/ecc.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    /// Witnesses the given scalar as a private input to the circuit.
+    fn witness_scalar(
+        layouter: &mut impl Layouter<Self>,
+        value: impl FnMut() -> Result<C::Scalar, Error>,
+    ) -> Result<Self::Scalar, Error>;
+
+    /// Witnesses the given point as a private input to the circuit.
+    fn witness_point(
+        layouter: &mut impl Layouter<Self>,
+        value: impl FnMut() -> Result<C, Error>,
```
Comment (str4d) [inline src/gadget/ecc.rs:29]:
```suggestion
        value: Option<C>,
```

File: src/gadget/ecc.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    ) -> Result<Self::Point, Error>;
+
+    /// Exposes the given point as a public input to the circuit.
+    fn expose_point(layouter: &mut impl Layouter<Self>, a: &Self::Point) -> Result<(), Error>;
+
+    /// Loads a fixed point into the circuit.
+    fn load_fixed(
+        layouter: &mut impl Layouter<Self>,
+        value: impl FnMut() -> Result<C, Error>,
```
Comment (str4d) [inline src/gadget/ecc.rs:38]:
```suggestion
        value: Option<C>,
```

**Review by ebfull — APPROVED** [review]

**Pull request merged**